### ISSUE TITLE
How to troubleshoot ALPN check failure?

### ISSUE THREAD

**Issue by davidvartan** [grpc/grpc-swift#278]
Using 240466e249330bde82395b027e29f1f438221cea

I'm setting up a gRPC connection to a backend with Envoy proxy doing TLS termination. Cert is from LetsEncrypt.

Was seeing the connection being terminated during the TLS handshake:

> D0801 00:21:48.452617000 123145539833856 security_handshaker.cc:129]   Security handshake failed: {"created":"@1533050508.452603000","description":"Cannot check peer: missing selected ALPN property.","file":"/(edited)/grpc-swift/Sources/CgRPC/src/core/lib/security/security_connector/security_connector.cc","file_line":858}

Commenting out this bit of code makes things work fine: https://github.com/grpc/grpc-swift/blob/master/Sources/CgRPC/src/core/lib/security/security_connector/security_connector.cc#L852-L862

I also saw something about toggling `TSI_OPENSSL_ALPN_SUPPORT` in `vendor-grpc.sh` and not sure if ALPN is meant to be supported (I'm on Mac.)

Is there any guidance on troubleshooting this? Obviously, the problem could also be on Envoy's end, but it seems like diagnostic tools (`curl`, `openssl`, etc.) are able to negotiate.

Thank you!

`curl -v --http2` excerpt:

```
* TCP_NODELAY set
* Connected to <redacted> port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2

```

`openssl s_client -alpn h2 -connect` excerpt:

```
ALPN protocol: h2
```

**Comment by MrMage** [comment]
On Mac we are using BoringSSL, which would probably support ALPN. Could you try enabling that by changing https://github.com/grpc/grpc-swift/blob/240466e249330bde82395b027e29f1f438221cea/Sources/CgRPC/src/core/tsi/ssl_transport_security.cc#L65 and let me know if that helps?

**Comment by davidvartan** [comment]
Thanks, this appears to work!

With `TSI_OPENSSL_ALPN_SUPPORT 0`, the connection works fine for `datastore.googleapis.com` but fails for my (ALPN-enabled) Envoy server on that check in `security_connector.cc`.

With `TSI_OPENSSL_ALPN_SUPPORT 1` can connect fine to my server.

So there is something unique when (server is Envoy) && (ALPN is disabled in CgRPC source). Wondering what is the correct resolution here / especially if anyone else runs into this.

- Should the check in `security_connector.cc` be surrounded with `#if TSI_OPENSSL_ALPN_SUPPORT`? This is what I originally did to unblock myself, but then again, likely the real bug may be somewhere else because `datastore.googleapis.com` connects fine.

My uneducated guess is that in the first step of the handshake (https://cdn.keycdn.com/support/wp-content/uploads/2015/12/how-alpn-works.webp) CgRPC sends the request without ALPN (it's disabled in the header), and Envoy and Google respond slightly differently here. Question is if it's on CgRPC to be more resilient after that, or if if Envoy's response not up to spec.

**Comment by MrMage** [comment]
Sorry, no idea about this. If `security_connector.cc` were to be changed, that would need to happen in the upstream https://github.com/grpc/grpc repo anyway. I'm using Google's Endpoints Service Proxy, which supports HTTP1.1 and HTTP2 over SSL on the same port, though, and they seem to work even with OpenSSL ALPN disabled in my client (i.e. I didn't take the steps I described before).

**Comment by davidvartan** [comment]
Thanks! Yeah `TSI_OPENSSL_ALPN_SUPPORT 1` solves the immediate problem, but things should work correctly even with `0`. I'm just not sure which side (CgRPC or Envoy) is mishandling the handshake. Sounds like when I have more bandwidth I'll need to set up a repro and follow up with both projects.

**Comment by MrMage** [comment]
Could this be related? https://github.com/envoyproxy/envoy/issues/3394

(Again, I'm just fishing in the dark, don't know much about this stuff.)

**Comment by MrMage** [comment]
Closing this now; please open a new issue if the issue persists in Swift gRPC 0.9.0.
